src/docfx/build/xref/XrefMapLoader.cs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Text.Json;
+
+namespace Microsoft.Docs.Build
+{
+    internal class XrefMapLoader
+    {
+        private static byte[] s_uidBytes = Encoding.UTF8.GetBytes("uid");
+
+        public static ListBuilder<(string, Lazy<IXrefSpec>)> Load(string filePath)
+        {
+            var result = new ListBuilder<(string, Lazy<IXrefSpec>)>();
+            var content = File.ReadAllBytes(filePath);
+
+            // TODO: cache this position mapping if xref map file not updated, reuse it
+            var xrefSpecPositions = GetXrefSpecPositions(content);
+
+            foreach (var (uid, start, end) in xrefSpecPositions)
+            {
+                result.Add((uid, new Lazy<IXrefSpec>(() =>
+                {
+                    using (var stream = File.Open(filePath, FileMode.Open, FileAccess.Read, FileShare.Read))
+                    {
+                        var json = ReadJsonFragment(stream, start, end);
+                        return JsonUtility.Deserialize<ExternalXrefSpec>(json, filePath);
+                    }
+                })));
+            }
+            return result;
+        }
+
+        private static List<(string uid, long start, long end)> GetXrefSpecPositions(ReadOnlySpan<byte> content)
+        {
+            var result = new List<(string uid, long start, long end)>();
+            var reader = new Utf8JsonReader(content, isFinalBlock: true, default);
+            string uid = null;
+            int startIndex = 0;
+            int endIndex = 0;
+            while (reader.Read())
+            {
+                switch (reader.TokenType)
+                {
+                    case JsonTokenType.PropertyName:
+                        if (reader.TextEquals(s_uidBytes) && reader.Read() && reader.TokenType == JsonTokenType.String)
+                        {
+                            uid = Encoding.UTF8.GetString(reader.ValueSpan);
+                        }
+                        break;
+                    case JsonTokenType.StartObject:
+                        startIndex = (int)reader.TokenStartIndex;
```

Comment:
this needs to be pushed to a stack for cases like, add a test case for it.
```json
{
 "uid": "a",
  "prop": { }
}
```

Reply:
I thought this is not a valid case for the current `XrefSpec` model, but if xref property can be an JObject, sure...